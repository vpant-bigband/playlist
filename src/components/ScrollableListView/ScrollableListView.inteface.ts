export type eventType = 'window' | 'base';

export interface IScrollableListView {
  children?: React.ReactChildren | any[] | React.ReactNode;
  onScrollEndHandler?: () => void;
  renderRow?: (item: any, index: number) => React.ReactElement | React.ReactElement[] | any[] | null | string;
  offset?: number;
  data?: any[];
  testId?: string;
  loader?: React.ReactElement | React.ReactElement[] | any[] | null | string;
  attachEvent?: string;
  hasMore?: boolean;
}
