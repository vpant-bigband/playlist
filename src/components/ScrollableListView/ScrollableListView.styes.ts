import styled from '@emotion/styled';

export const ScrollableListViewWrapper = styled.ul`
  display: block;
  overflow: auto;
  flex: auto;
  width: 100%;
  list-style: none;
  margin: 0px;
  padding: 0px;
`;

export const RowItemWrapper = styled.div<{ height: number }>`
  min-height: ${(props) => (props.height > 0 ? props.height : 100)}px;
`;
