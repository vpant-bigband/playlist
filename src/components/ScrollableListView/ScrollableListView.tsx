import React, { useRef, useEffect, useState } from 'react';

import { ScrollableListViewWrapper, RowItemWrapper } from './ScrollableListView.styes';
import { IScrollableListView } from './ScrollableListView.inteface';

const ScrollableListView: React.FunctionComponent<IScrollableListView> = ({
  onScrollEndHandler,
  testId = 'scrollable-list-view',
  hasMore = true,
  children,
}) => {
  const loader = useRef(onScrollEndHandler);
  const [element, setElement] = useState<HTMLElement | null>(null);

  const observer = useRef(
    new IntersectionObserver(
      (entries) => {
        const first = entries[0];
        if (first.isIntersecting && loader.current) {
          loader.current();
        }
      },
      { threshold: 0.25 },
    ),
  );

  useEffect(() => {
    loader.current = onScrollEndHandler;
  }, [onScrollEndHandler]);

  useEffect(() => {
    const currentElement = element;
    const currentObserver = observer.current;
    if (currentElement) {
      currentObserver.observe(currentElement);
    }

    return () => {
      if (currentElement) {
        currentObserver.unobserve(currentElement);
      }
    };
  }, [element]);

  if (!children) {
    return null;
  }
  return (
    <ScrollableListViewWrapper data-testid={testId}>
      {React.Children.map(children, (child, rowIndex: number) => {
        return <RowItem key={rowIndex}>{child}</RowItem>;
      })}
      {hasMore && <li data-testid="load-more-element" ref={setElement} />}
    </ScrollableListViewWrapper>
  );
};

const RowItem: React.FunctionComponent<{
  children: React.ReactChild | string | null | any;
}> = ({ children }) => {
  //@ts-ignore
  const [inView, setInView] = useState<boolean>( window.isIntesecting || false);
  const [height, setHeight] = useState<number>(0);
  const [element, setElement] = useState<HTMLElement | null>(null);
  const observer = useRef(
    new IntersectionObserver(
      (entries) => {
        const first = entries[0];
        if (first.isIntersecting) {
          setInView(true);
        }
      },
      { threshold: 0.50 },
    ),
  );
  useEffect(() => {
    const currentElement = element;
    const currentObserver = observer.current;
    if (currentElement) {
      currentObserver.observe(currentElement);
      // @ts-ignore
      const childHeight = currentElement.firstChild ? currentElement.firstChild.clientHeight : 0;
      if (childHeight > 0) {
        setHeight((prevHeight) => (prevHeight !== childHeight ? childHeight : prevHeight));
      }
    }
    return () => {
      if (currentElement) {
        currentObserver.unobserve(currentElement);
      }
    };
  }, [element]);
  return (
    <RowItemWrapper height={height} ref={setElement}>
      {inView && children}
    </RowItemWrapper>
  );
};

export default ScrollableListView;
