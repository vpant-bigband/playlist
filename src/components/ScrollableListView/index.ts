import ScrollableListView from './ScrollableListView';
export default ScrollableListView;
