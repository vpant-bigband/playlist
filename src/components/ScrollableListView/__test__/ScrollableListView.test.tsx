import React from 'react';
import { mockAllIsIntersecting } from 'react-intersection-observer/test-utils';
import { render } from '@testing-library/react';
import ScrollableListView from '../';
import data from '../__mocks/mockListData';


jest.clearAllMocks();

const observe = jest.fn();
const unobserve = jest.fn();

// @ts-ignore
window.IntersectionObserver = jest.fn(() => ({
  observe,
  unobserve,
}));

const renderRow = ({ id, name }: { id: string; name: string }) => {
  return (
    <li style={{ height: 150 }} key={id}>
      {name}
    </li>
  );
};

describe('ScrollableListView test suite', () => {
  beforeEach(() => {
    mockAllIsIntersecting(true);
  });
  it('should be match snapshot', () => {
    const { container } = render(<ScrollableListView>{data[0].map(renderRow)}</ScrollableListView>);
    expect(container).toMatchSnapshot();
  });
  it('should invoke window.IntersectionObserver method', () => {
    render(<ScrollableListView>{data[0].map(renderRow)}</ScrollableListView>);
    expect(window.IntersectionObserver).toHaveBeenCalled();
  });
  it('should invoke onScrollEndHandler when scroll the end', async () => {
    const onScrollEndHandlerSpy = jest.fn();
    const observeFn = jest.fn((elementNode: any) => {
      onScrollEndHandlerSpy(elementNode);
    });
    // @ts-ignore
    window.IntersectionObserver = jest.fn(() => ({
      observe: observeFn,
      unobserve,
    }));
    const { findByTestId } = render(
      <ScrollableListView onScrollEndHandler={onScrollEndHandlerSpy}>{[renderRow(data[0][0])]}</ScrollableListView>,
    );
    const element = await findByTestId('scrollable-list-view');
    expect(element).toBeDefined();
    const loadMore = await findByTestId('load-more-element');
    expect(loadMore).toBeInTheDocument();
    expect(observeFn).toHaveBeenCalled();
  });
});
