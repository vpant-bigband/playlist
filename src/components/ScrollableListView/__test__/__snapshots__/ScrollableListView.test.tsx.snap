// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollableListView test suite should be match snapshot 1`] = `
<div>
  <ul
    class="css-jdcg90"
    data-testid="scrollable-list-view"
  >
    <div
      class="css-cphitw"
      height="0"
    />
    <div
      class="css-cphitw"
      height="0"
    />
    <div
      class="css-cphitw"
      height="0"
    />
    <div
      class="css-cphitw"
      height="0"
    />
    <div
      class="css-cphitw"
      height="0"
    />
    <div
      class="css-cphitw"
      height="0"
    />
    <div
      class="css-cphitw"
      height="0"
    />
    <div
      class="css-cphitw"
      height="0"
    />
    <div
      class="css-cphitw"
      height="0"
    />
    <li
      data-testid="load-more-element"
    />
  </ul>
</div>
`;
